superduper/jobs/queue.py

```diff
@@ -154,34 +154,26 @@ class LocalQueueConsumer(BaseQueueConsumer):
     def start_consuming(self):
         """Start consuming."""
 
+    def _get_consumers(self, db, components):
+        components = list(set(components.keys()))
+        components_to_use = []
+
+        for type_id, _ in components:
+            components_to_use += [(type_id, x) for x in db.show(type_id)]
+        return set(components_to_use + components)
+
     def consume(self, db: 'Datalayer', queue: t.Dict, components: t.Dict):
         """Consume the current queue and run jobs."""
-        from superduper.base.datalayer import DBEvent
-
         queue_jobs = defaultdict(lambda: [])
-        components_to_use = [('listener', x) for x in db.show('listener')]
-        components_to_use += [('vector_index', x) for x in db.show('vector_index')]
-        for type_id, identifier in components_to_use:
-            events = queue[type_id, identifier]
-            if not events:
-                continue
-            queue[type_id, identifier] = []
-            component = components[type_id, identifier]
-            jobs = []
-            for event_type, type_events in DBEvent.chunk_by_event(events).items():
-                ids = [event.id for event in type_events]
-                overwrite = (
-                    True if event_type in [DBEvent.insert, DBEvent.upsert] else False
-                )
-                logging.info(
-                    f'Running jobs for {component.type_id}::{component.identifier}'
-                )
-                logging.debug(f'Using ids: {ids}')
-                job = component.run_jobs(
-                    db=db, ids=ids, overwrite=overwrite, event_type=event_type
-                )
-                jobs.append(job)
-            queue_jobs[type_id, identifier].extend(jobs)
+        consumers = self._get_consumers(db, components)
+        # All consumer are executed one by one.
+        for consumer in consumers:
+            events = queue[consumer]
+            queue[consumer] = []
+            component = components[consumer]
+            # Consume
+            jobs = consume_events(db, component=component, events=events)
+            queue_jobs[consumer].extend(jobs)
```

Comment:
I’m not sure if this will break the dependencies or the running order

Reply:
fixed